### ISSUE TITLE
"Request codeLens/resolve failed" error when using collection expression spread syntax

### ISSUE THREAD

**Issue by aelij** [dotnet/roslyn#71339]
**Version Used**: C# 12, VS Code C# extension v2.14.8, C# Dev Kit v1.2.5

Does not happen in VS 17.9.0 preview 2.

**Steps to Reproduce**:

Create a project with the following code and open in VS Code:

```cs
class Test
{

  public Dictionary<string, object> Config => [
    .. GetConfig(),
  ];

  private Dictionary<string, object> GetConfig() => new();
}

static class DictionaryExtensions
{
  public static void Add<TKey, TValue>(this IDictionary<TKey, TValue> dictionary, KeyValuePair<TKey, TValue> keyValuePair)
  {
    dictionary.Add(keyValuePair.Key, keyValuePair.Value);
  }
}
```

**Diagnostic Id**: N/A

**Expected Behavior**: No error

**Actual Behavior**: 

Notification with the following error:
> Request codeLens/resolve failed.

Log:

```
[Error - 2:22:23 PM] [LanguageServerHost] System.InvalidOperationException: Unexpected value 'Microsoft.CodeAnalysis.CSharp.BoundCall' of type 'Microsoft.CodeAnalysis.CSharp.BoundCall'
   at Microsoft.CodeAnalysis.CSharp.BoundNode.GetConversion(BoundExpression conversion, BoundValuePlaceholder placeholder) in /_/src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs:line 452
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundCollectionExpressionSpreadElement(BoundCollectionExpressionSpreadElement element) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 1291
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundCollectionExpressionElement(BoundNode element) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 1267
   at System.Collections.Immutable.ImmutableArray.CreateRange[TSource,TResult](ImmutableArray`1 items, Func`2 selector)
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundCollectionExpression(BoundCollectionExpression expr) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 1230
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 335
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundConversionOperation(BoundConversion boundConversion, Boolean forceOperandImplicitLiteral) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 1098
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 335
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundReturnStatementOperation(BoundReturnStatement boundReturnStatement) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 2047
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 335
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateFromArray[TBoundNode,TOperation](ImmutableArray`1 boundNodes) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 341
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.CreateBoundBlockOperation(BoundBlock boundBlock) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 1773
   at Microsoft.CodeAnalysis.Operations.CSharpOperationFactory.Create(BoundNode boundNode) in /_/src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:line 335
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetRootOperation() in /_/src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:line 1191
   at Microsoft.CodeAnalysis.CSharp.MemberSemanticModel.GetOperationWorker(CSharpSyntaxNode node, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs:line 1173
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetOperationWorker(CSharpSyntaxNode node, CancellationToken cancellationToken) in /_/src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:line 178
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.GetOperationBlocksToAnalyze(ImmutableArray`1 executableBlocks, SemanticModel semanticModel, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 2927
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.<>c__DisplayClass11_0.<ExecuteDeclaringReferenceActions>g__executeExecutableCodeActions|2() in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 2711
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SyntaxReference decl, SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, GroupedAnalyzerActions coreActions, GroupedAnalyzerActions additionalPerSymbolActions, Boolean shouldExecuteSyntaxNodeActions, Boolean shouldExecuteOperationActions, Boolean shouldExecuteCodeBlockActions, Boolean shouldExecuteOperationBlockActions, Boolean isInGeneratedCode, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, Boolean isGeneratedCodeSymbol, IGroupedAnalyzerActions additionalPerSymbolActions, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 2526
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.TryProcessSymbolDeclaredAsync(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 1781
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.TryProcessEventCoreAsync(CompilationEvent compilationEvent, AnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 1739
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessEventAsync(CompilationEvent e, AnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 1651
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessCompilationEventsCoreAsync(AnalysisScope analysisScope, Boolean prePopulatedEventQueue, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 1629
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessCompilationEventsAsync(AnalysisScope analysisScope, Boolean prePopulatedEventQueue, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 1570
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ExecutePrimaryAnalysisTaskAsync(AnalysisScope analysisScope, Boolean usingPrePopulatedEventQueue, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 714
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.AttachQueueAndProcessAllEventsAsync(AsyncQueue`1 eventQueue, AnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs:line 654
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.<ComputeAnalyzerDiagnosticsAsync>g__attachQueueAndProcessAllEventsAsync|62_0(ArrayBuilder`1 builder, AnalyzerDriver driver, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:line 827
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.ComputeAnalyzerDiagnosticsAsync(AnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:line 967
   at Microsoft.CodeAnalysis.Diagnostics.CompilationWithAnalyzers.GetAnalysisResultCoreAsync(SemanticModel model, Nullable`1 filterSpan, ImmutableArray`1 analyzers, CancellationToken cancellationToken) in /_/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:line 655
   at Microsoft.CodeAnalysis.Diagnostics.Extensions.GetAnalysisResultAsync(CompilationWithAnalyzers compilationWithAnalyzers, DocumentAnalysisScope documentAnalysisScope, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Diagnostics/Extensions.cs:line 363
   at Microsoft.CodeAnalysis.Diagnostics.Extensions.GetAnalysisResultAsync(CompilationWithAnalyzers compilationWithAnalyzers, DocumentAnalysisScope documentAnalysisScope, Project project, DiagnosticAnalyzerInfoCache analyzerInfoCache, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Diagnostics/Extensions.cs:line 329
   at Microsoft.CodeAnalysis.Diagnostics.InProcOrRemoteHostAnalyzerRunner.AnalyzeInProcAsync(DocumentAnalysisScope documentAnalysisScope, Project project, CompilationWithAnalyzers compilationWithAnalyzers, RemoteHostClient client, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs:line 148
   at Microsoft.CodeAnalysis.Diagnostics.InProcOrRemoteHostAnalyzerRunner.<>c__DisplayClass7_0.<<AnalyzeAsync>g__AnalyzeCoreAsync|0>d.MoveNext() in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs:line 87
--- End of stack trace from previous location ---
   at Microsoft.CodeAnalysis.Diagnostics.InProcOrRemoteHostAnalyzerRunner.AnalyzeAsync(DocumentAnalysisScope documentAnalysisScope, Project project, CompilationWithAnalyzers compilationWithAnalyzers, Boolean isExplicit, Boolean forceExecuteAllAnalyzers, Boolean logPerformanceInfo, Boolean getTelemetryInfo, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.InProcOrRemoteHostAnalyzerRunner.cs:line 88
   at Microsoft.CodeAnalysis.Diagnostics.DocumentAnalysisExecutor.GetAnalysisResultAsync(DocumentAnalysisScope analysisScope, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/DocumentAnalysisExecutor.cs:line 197
   at Microsoft.CodeAnalysis.Diagnostics.DocumentAnalysisExecutor.GetSemanticDiagnosticsAsync(DiagnosticAnalyzer analyzer, Boolean isCompilerAnalyzer, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/DocumentAnalysisExecutor.cs:line 386
   at Microsoft.CodeAnalysis.Diagnostics.DocumentAnalysisExecutor.ComputeDiagnosticsAsync(DiagnosticAnalyzer analyzer, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/DocumentAnalysisExecutor.cs:line 180
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.LatestDiagnosticsForSpanGetter.ComputeDocumentDiagnosticsForAnalyzerCoreAsync(DiagnosticAnalyzer analyzer, DocumentAnalysisExecutor executor, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 577
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.LatestDiagnosticsForSpanGetter.ComputeDocumentDiagnosticsCoreAsync(DocumentAnalysisExecutor executor, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 559
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.LatestDiagnosticsForSpanGetter.ComputeDocumentDiagnosticsAsync(ImmutableArray`1 analyzersWithState, AnalysisKind kind, Nullable`1 span, ArrayBuilder`1 builder, Boolean incrementalAnalysis, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 545
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.LatestDiagnosticsForSpanGetter.TryGetAsync(ArrayBuilder`1 list, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 360
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.TryAppendDiagnosticsForSpanAsync(TextDocument document, Nullable`1 range, ArrayBuilder`1 result, Func`2 shouldIncludeDiagnostic, Boolean includeSuppressedDiagnostics, Boolean includeCompilerDiagnostics, ICodeActionRequestPriorityProvider priorityProvider, Boolean blockForData, Func`2 addOperationScope, DiagnosticKind diagnosticKinds, Boolean isExplicit, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 35
   at Microsoft.CodeAnalysis.Diagnostics.EngineV2.DiagnosticIncrementalAnalyzer.GetDiagnosticsForSpanAsync(TextDocument document, Nullable`1 range, Func`2 shouldIncludeDiagnostic, Boolean includeSuppressedDiagnostics, Boolean includeCompilerDiagnostics, ICodeActionRequestPriorityProvider priorityProvider, Boolean blockForData, Func`2 addOperationScope, DiagnosticKind diagnosticKinds, Boolean isExplicit, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnosticsForSpan.cs:line 56
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.StreamFixesAsync(TextDocument document, TextSpan range, ICodeActionRequestPriorityProvider priorityProvider, CodeActionOptionsProvider fallbackOptions, Func`2 addOperationScope, CancellationToken cancellationToken)+MoveNext() in /_/src/Features/LanguageServer/Protocol/Features/CodeFixes/CodeFixService.cs:line 242
   at Microsoft.CodeAnalysis.CodeFixes.CodeFixService.StreamFixesAsync(TextDocument document, TextSpan range, ICodeActionRequestPriorityProvider priorityProvider, CodeActionOptionsProvider fallbackOptions, Func`2 addOperationScope, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Microsoft.CodeAnalysis.Shared.Extensions.IAsyncEnumerableExtensions.ToImmutableArrayAsync[T](IAsyncEnumerable`1 values, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Shared/Extensions/IAsyncEnumerableExtensions.cs:line 36
   at Microsoft.CodeAnalysis.Shared.Extensions.IAsyncEnumerableExtensions.ToImmutableArrayAsync[T](IAsyncEnumerable`1 values, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Shared/Extensions/IAsyncEnumerableExtensions.cs:line 36
   at Microsoft.CodeAnalysis.UnifiedSuggestions.UnifiedSuggestedActionsSource.GetFilterAndOrderCodeFixesAsync(Workspace workspace, ICodeFixService codeFixService, TextDocument document, TextSpan selection, ICodeActionRequestPriorityProvider priorityProvider, CodeActionOptionsProvider fallbackOptions, Func`2 addOperationScope, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Features/UnifiedSuggestions/UnifiedSuggestedActionsSource.cs:line 61
   at Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions.CodeActionHelpers.GetActionSetsAsync(Document document, CodeActionOptionsProvider fallbackOptions, ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, Range selection, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:line 417
   at Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActions.CodeActionHelpers.GetVSCodeActionsAsync(CodeActionParams request, Document document, CodeActionOptionsProvider fallbackOptions, ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, Boolean hasVsLspCapability, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs:line 93
   at Microsoft.CodeAnalysis.LanguageServer.Handler.CodeActionsHandler.HandleRequestAsync(CodeActionParams request, RequestContext context, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs:line 67
   at Microsoft.CommonLanguageServerProtocol.Framework.QueueItem`3.StartRequestAsync(TRequestContext context, CancellationToken cancellationToken) in /_/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/QueueItem.cs:line 176
[Error - 2:22:23 PM] Request textDocument/codeAction failed.
  Message: Unexpected value 'Microsoft.CodeAnalysis.CSharp.BoundCall' of type 'Microsoft.CodeAnalysis.CSharp.BoundCall'
  Code: -32000 
```

**Comment by CyrusNajmabadi** [comment]
@cston Looks like a IOp bug with collection expressions.

**Comment by cston** [comment]
@aelij, thanks for reporting this issue.

I cannot repro the exception with the exact code from the description, but the exception is thrown when the extension method is unavailable or inapplicable. Perhaps the exception occurred while editing the code from the description.

For instance, the issue repros with the following where there is no extension method. 
```csharp
class Test
{
    Dictionary<string, object> Config => [
        .. GetConfig(),
    ];
    private Dictionary<string, object> GetConfig() => new();
}
```

**Comment by cston** [comment]
Fixed in https://github.com/dotnet/roslyn/issues/71339.
